modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/DefaultExecutionMediator.java

```diff
@@ -243,7 +243,109 @@ private void intercept(List<Interceptor> interceptors, Consumer<Interceptor> clo
         });
     }
 
-    private Throwable handleException(Exception e)
+    private class OperationRetryCallBack implements RetryCallback
+    {
+
+        private OperationExecutor operationExecutor;
+        private final OperationContext context;
+        private final List<Interceptor> interceptorList;
+        private OperationExecutionResult operationExecutionResult;
+
+        private OperationRetryCallBack(OperationExecutor operationExecutor, OperationContext context, List<Interceptor> interceptorList)
+        {
+
+            this.operationExecutor = operationExecutor;
+            this.context = context;
+            this.interceptorList = interceptorList;
+        }
+
+        @Override
+        public void doWork(RetryContext retryContext) throws Exception
+        {
+            operationExecutionResult = executeWithInterceptors(operationExecutor, context, interceptorList, new ValueHolder<>());
+
+            if (!operationExecutionResult.isOk())
+            {
+                if (operationExecutionResult.getException() instanceof ConnectionException || (operationExecutionResult.getRetryRequest().isPresent() && operationExecutionResult.getRetryRequest().get().isRetryRequested()))
+                {
+                    throw (Exception) operationExecutionResult.getException();
+                }
+                else
+                {
+                    retryContext.setFailed(operationExecutionResult.getException());
+                }
+            }
+
+        }
+
+        @Override
+        public String getWorkDescription()
+        {
+            return String.format("Extension [%s] with configuration [%s]", context.getConfiguration().getModel().getExtensionModel().getName(), context.getConfiguration().getName());
+        }
+
+        @Override
+        public Object getWorkOwner()
+        {
+            return this;
+        }
+
+        public OperationExecutionResult getOperationExecutionResult()
+        {
+            return operationExecutionResult;
+        }
+    }
+
+    private RetryPolicyTemplate getRetryPolicyTemplateFromConnectionProvider(Optional<ConnectionProvider> connectionProvider)
+    {
+        RetryPolicyTemplate retryPolicyTemplate;
+        if (connectionProvider.isPresent() && ConnectionProviderWrapper.class.isAssignableFrom(connectionProvider.get().getClass()))
```

Comment:
please tweak this to be more readable... if connectionProvider is not present then quickly return the default one.
Also change the code so that connectionProvider.get() is invoked only once.